### PR TITLE
Address UIDelays in AbstractGraphProvider.GetCommands

### PR DESCRIPTION
**Pull request by ToddGrun**
The .Any calls in this method are showing up in prism uidelays (not particularly high though as there are only 124 expressed hits in 17.8). This PR reduces some duplicated calls in this method, reducing the number of linear searches through nodes in about half in most codepaths.

I don't have much context on this code, so I'm also looking into whether both the VB and C# derivations of this class should be getting called all the time, but any changes in that regard would be separate from this.

Partially addresses [AB#1936654](https://devdiv.visualstudio.com/0bdbc590-a062-4c3f-b0f6-9383f67865ee/_workitems/edit/1936654)

### CODE REVIEW

**Comment by sharwell** [comment]
I do not believe this change will have any impact on the indicated issue. PRISM indicates the delays occur during JIT operations, and these operations would always occur on the first call of a sequence.

**Comment by ToddGrun** [comment]
> I do not believe this change will have any impact on the indicated issue. PRISM indicates the delays occur during JIT operations, and these operations would always occur on the first call of a sequence.

I'd love to understand this better. Are you saying that only the First of the duplicated .Any calls in the method would incur the overhead shown in prism? I definitely don't understand what clr.dll!StubDispatchFixupStub is doing, but it doesn't show the standard clrjit method calls that I associate with jit costs.

**Comment by sharwell** [comment]
> Are you saying that only the First of the duplicated .Any calls in the method would incur the overhead shown in prism?

That's my interpretation of what I was seeing, but it depends a bit on which part of the JIT is getting hit.

**Comment by ToddGrun** [comment]
> Are you saying that only the First of the duplicated .Any calls in the method would incur the overhead shown in prism?

Can you maybe elaborate a bit on which method you are correlating with jitting? I see some code that looks to be resolving the method as the primary cost under the .Any call, and I guess I'd like to understand how that's limited to the first .Any invocation.

**Comment by ToddGrun** [comment]
I tried measuring the relative impact of this change by running both old/new code 1000 times for each call into GetCommands, expanded a couple items in solution explorer, and inspected the relative timing differences (done against retail bits, no debugger attached until I needed to get the stopwatch timings)

image

**Comment by CyrusNajmabadi** [comment]
Wow.  That's still 5s of CPU.  How large are these collections?

**Review by CyrusNajmabadi — APPROVED** [review]

File: src/VisualStudio/Core/Def/Progression/GraphProvider.cs [hunk 1]
```diff
@@ -176,10 +176,12 @@ public IEnumerable<GraphCommand> GetCommands(IEnumerable<GraphNode> nodes)
     {
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Progression/GraphProvider.cs:176]: really dumb question.  is the IENumerable passed in here lazy?  If so, can we def make it *non* lazy (an immutable-array would be good).  I would hate if each .Any was lazily evaluating that underlying list again and again.

Comment (CyrusNajmabadi) [reply src/VisualStudio/Core/Def/Progression/GraphProvider.cs:176]: oh.  we don't even own this list.  

*consider* acutally making a single loop that assigns into locals (or a helper function for this).  having one iteration would make me feel a lot more confident about this.

**Review by CyrusNajmabadi — APPROVED** [review]
approved just for the benefit.  but some ideas to see if we can make things better.  I'd love t know how big this list is.  and we likely shoudl converse with progression team about getting this off the UI thread entirely.

**Comment by ToddGrun** [comment]
> approved just for the benefit. but some ideas to see if we can make things better. I'd love t know how big this list is. and we likely shoudl converse with progression team about getting this off the UI thread entirely.

As mentioned offline, I don't want approval for a benefit that was perceived to be larger than it is. I don't see this method in my local profiles taking seconds (the numbers from the image above were generated based on calling into both the old and new implementation 1000 times), but there is data on prism indicating it is costly for some users. I haven't thought about your comments above and will comment on them later tonight or tomorrow morning.

File: src/VisualStudio/Core/Def/Progression/GraphProvider.cs (hunk 1 above)
Comment (ToddGrun) [reply src/VisualStudio/Core/Def/Progression/GraphProvider.cs:176]: OK, I modified the code to do a single loop for most of the data needed. I didn't have that loop calculate all the data as some of it is only needed for certain symbolkind/typekind combinations. The new (artificial) numbers still look good:

```
s_swOld.ElapsedMilliseconds	1939	long
s_swNew.ElapsedMilliseconds	426	long
```

File: src/VisualStudio/Core/Def/Progression/GraphProvider.cs [hunk 2]
```diff
@@ -176,10 +176,13 @@ public IEnumerable<GraphCommand> GetCommands(IEnumerable<GraphNode> nodes)
...
 
+        // Gather most information about the nodes in a single pass.
+        var nodesInfo = GetTypeKindAndSymbolKindInfo(nodes);
+
         // Only nodes that explicitly state that they contain children (e.g., source files) and named types should
         // be expandable.
-        if (nodes.Any(n => n.Properties.Any(p => p.Key == DgmlNodeProperties.ContainsChildren)) ||
-            nodes.Any(n => IsAnySymbolKind(n, SymbolKind.NamedType)))
+        if (nodesInfo.ContainsSymbolKindNamedType ||
+            nodes.Any(static n => n.Properties.Any(static p => p.Key == DgmlNodeProperties.ContainsChildren)))
```
Comment (sharwell) [inline src/VisualStudio/Core/Def/Progression/GraphProvider.cs:185]: 😕 Can you not just use `(bool?)n[DgmlNodeProperties.ContainsChildren] ?? false` or something like that? Currently this code appears to assume that if the key exists, its value is true. I'm not sure that's valid.

File: src/VisualStudio/Core/Def/Progression/GraphProvider.cs [hunk 3]
```diff
@@ -222,8 +225,8 @@ public IEnumerable<GraphCommand> GetCommands(IEnumerable<GraphNode> nodes)
         }
 
         // Only show 'Is Called By' on an applicable member in a class or struct
-        if (nodes.Any(n => IsAnySymbolKind(n, SymbolKind.Event, SymbolKind.Method, SymbolKind.Property) &&
-                           IsAnyTypeKind(n, TypeKind.Class, TypeKind.Struct)))
+        if (nodesInfo.ContainsSymbolKindEventMethodOrProperty &&
+            nodesInfo.ContainsTypeKindClassOrStruct)
```
Comment (sharwell) [inline src/VisualStudio/Core/Def/Progression/GraphProvider.cs:229]: ❗  This seems to be a different meaning. Are we sure this is correct?

**Review by sharwell — CHANGES_REQUESTED** [review]

File: src/VisualStudio/Core/Def/Progression/GraphProvider.cs [hunk 4]
```diff
@@ -189,14 +192,14 @@ public IEnumerable<GraphCommand> GetCommands(IEnumerable<GraphNode> nodes)
...
+        if (nodesInfo.ContainsNonRoslynNode)
         {
             yield break;
         }
 
         // Only show 'Base Types' and 'Derived Types' on a class or interface.
-        if (nodes.Any(n => IsAnySymbolKind(n, SymbolKind.NamedType) &&
-                           IsAnyTypeKind(n, TypeKind.Class, TypeKind.Interface, TypeKind.Struct, TypeKind.Enum, TypeKind.Delegate)))
+        if (nodesInfo.ContainsSymbolKindNamedType &&
+            (nodesInfo.ContainsTypeKindClassOrStruct || nodesInfo.ContainsTypeKindInterface || nodesInfo.ContainsTypeKindEnumOrDelegate))
```
Comment (sharwell) [inline src/VisualStudio/Core/Def/Progression/GraphProvider.cs:202]: 📝 The associativity of `&&` here seems to have changed

File: src/VisualStudio/Core/Def/Progression/GraphProvider.cs [hunk 5]
```diff
@@ -252,12 +255,12 @@ public IEnumerable<GraphCommand> GetCommands(IEnumerable<GraphNode> nodes)
 
         // Show 'Implements' on public, non-static members of a class or struct.  Note: we should
         // also show it on explicit interface impls in C#.
-        if (nodes.Any(n => IsAnySymbolKind(n, SymbolKind.Event, SymbolKind.Method, SymbolKind.Property) &&
-                           IsAnyTypeKind(n, TypeKind.Class, TypeKind.Struct) &&
-                           !GetModifiers(n).IsStatic))
+        if (nodesInfo.ContainsSymbolKindEventMethodOrProperty &&
+            nodesInfo.ContainsTypeKindClassOrStruct &&
+            nodes.Any(static n => !GetModifiers(n).IsStatic))
```
Comment (sharwell) [inline src/VisualStudio/Core/Def/Progression/GraphProvider.cs:260]: 📝 The associativity of `&&` (both instances) here seems to have changed

File: src/VisualStudio/Core/Def/Progression/GraphProvider.cs [hunk 6]
```diff
@@ -268,8 +271,8 @@ public IEnumerable<GraphCommand> GetCommands(IEnumerable<GraphNode> nodes)
         }
 
         // Show 'Implemented By' on an interface.
-        if (nodes.Any(n => IsAnySymbolKind(n, SymbolKind.NamedType) &&
-                           IsAnyTypeKind(n, TypeKind.Interface)))
+        if (nodesInfo.ContainsSymbolKindNamedType &&
+            nodesInfo.ContainsTypeKindInterface)
```
Comment (sharwell) [inline src/VisualStudio/Core/Def/Progression/GraphProvider.cs:275]: 📝 The associativity of `&&` here seems to have changed

File: src/VisualStudio/Core/Def/Progression/GraphProvider.cs [hunk 7]
```diff
@@ -279,8 +282,8 @@ public IEnumerable<GraphCommand> GetCommands(IEnumerable<GraphNode> nodes)
         }
 
         // Show 'Implemented By' on any member of an interface.
-        if (nodes.Any(n => IsAnySymbolKind(n, SymbolKind.Event, SymbolKind.Method, SymbolKind.Property) &&
-                           IsAnyTypeKind(n, TypeKind.Interface)))
+        if (nodesInfo.ContainsSymbolKindEventMethodOrProperty &&
+            nodesInfo.ContainsTypeKindInterface)
```
Comment (sharwell) [inline src/VisualStudio/Core/Def/Progression/GraphProvider.cs:286]: 📝 The associativity of `&&` here seems to have changed

File: src/VisualStudio/Core/Def/Progression/GraphProvider.cs [hunk 8]
```diff
@@ -302,9 +305,9 @@ public IEnumerable<GraphCommand> GetCommands(IEnumerable<GraphNode> nodes)
         }
 
         // Show 'Overridden By' on any applicable member of a class or struct
-        if (nodes.Any(n => IsAnySymbolKind(n, SymbolKind.Event, SymbolKind.Method, SymbolKind.Property) &&
-                           IsAnyTypeKind(n, TypeKind.Class, TypeKind.Struct) &&
-                           IsOverridable(n)))
+        if (nodesInfo.ContainsSymbolKindEventMethodOrProperty &&
+            nodesInfo.ContainsTypeKindClassOrStruct &&
```
Comment (sharwell) [inline src/VisualStudio/Core/Def/Progression/GraphProvider.cs:309]: 📝 The associativity of `&&` here (both instances) seems to have changed

**Comment by ToddGrun** [comment]
>     public IEnumerable<GraphCommand> GetCommands(IEnumerable<GraphNode> nodes)

As an aside, this code currently gets called for both VB/C# providers. It's not trivial to filter on our side, and would probably end up being more complex and slower to do so.

Pinged the progression folks and notified them that from my point of view, the best way to address that would be on their side by either:

1) Allow us to register multiple project capabilities for a single provider (would allow us to only have a single provider that registers for both C#/VB)
2) Do the filtering of providers to call on their side before calling into them. They currently call all providers that are valid for the solution, not those that are valid for the project (which is the level at which we register)

---
Refers to: src/VisualStudio/Core/Def/Progression/GraphProvider.cs:175 in 1a1da36. [](commit_id = 1a1da36dd727e6623cf80983833d43201b6c7ee9, deletion_comment = False)

File: src/VisualStudio/Core/Def/Progression/GraphProvider.cs (hunk 2 above)
Comment (ToddGrun) [reply src/VisualStudio/Core/Def/Progression/GraphProvider.cs:185]: Maybe? I'm not sure as the semantics around properties is a bit odd (the GetValues call for properties filters properties that are valid based on the owner's schema). Either way, unless we're certain the current behavior isn't correct, I'd rather not change the behavior.

File: src/VisualStudio/Core/Def/Progression/GraphProvider.cs (hunk 4 above)
Comment (ToddGrun) [reply src/VisualStudio/Core/Def/Progression/GraphProvider.cs:202]: I'm not seeing the functional differences in the spots you've mentioned. Maybe an explicit example of an input that would have changed behavior would be the best way for me to understand your concern.

Comment (ToddGrun) [reply src/VisualStudio/Core/Def/Progression/GraphProvider.cs:202]: Oh, I see now. All callers I've encountered have called this with a single item in the array, but if there were multiple items, then these conditions are no longer such that a single item would have needed met the criteria.

**Pull request closed without merging**